mlir/lib/Catalyst/Driver/CompilerDriver.cpp

```diff
@@ -0,0 +1,334 @@
+    stablehlo::registerAllDialects(registry);
+
+    // Catalyst
+    registry.insert<CatalystDialect>();
+    registry.insert<quantum::QuantumDialect>();
+    registry.insert<gradient::GradientDialect>();
+}
+} // namespace
+
+template <typename MLIRObject> std::string serializeMLIRObject(MLIRObject &obj)
+{
+    std::string output;
+    llvm::raw_string_ostream ostream{output};
+    obj.print(ostream);
+    return output;
+}
+
+FailureOr<llvm::Function *> getJITFunction(MLIRContext *ctx, llvm::Module &llvmModule)
+{
+    Location loc = NameLoc::get(StringAttr::get(ctx, llvmModule.getName()));
+    for (auto &function : llvmModule.functions()) {
+        emitRemark(loc) << "Found LLVM function: " << function.getName() << "\n";
+        if (function.getName().starts_with("jit_")) {
+            return &function;
+        }
+    }
+    emitError(loc, "Failed to find JIT function");
+    return failure();
+}
+
+LogicalResult inferMLIRReturnTypes(MLIRContext *ctx, llvm::Type *returnType,
+                                   Type assumedElementType,
+                                   SmallVectorImpl<RankedTensorType> &inferredTypes)
+{
+    auto inferSingleMemRef = [&](llvm::StructType *descriptorType) {
+        SmallVector<int64_t> resultShape;
+        assert(descriptorType->getNumElements() >= 3 &&
+               "Expected MemRef descriptor struct to have at least 3 entries");
+        if (descriptorType->getNumElements() == 3) {
+            // resultShape is empty
+        }
+        else {
+            auto *arrayType = cast<llvm::ArrayType>(descriptorType->getTypeAtIndex(3));
+            size_t rank = arrayType->getNumElements();
+            for (size_t i = 0; i < rank; i++) {
+                resultShape.push_back(ShapedType::kDynamic);
+            }
+        };
+        return RankedTensorType::get(resultShape, assumedElementType);
+    };
+    if (returnType->isVoidTy()) {
+        return success();
+    }
+    if (auto *structType = dyn_cast<llvm::StructType>(returnType)) {
+        // The return type could be a single memref descriptor or a struct of multiple memref
+        // descriptors.
+        if (isa<llvm::StructType>(structType->getElementType(0))) {
+            for (size_t i = 0; i < structType->getNumElements(); i++) {
+                inferredTypes.push_back(
+                    inferSingleMemRef(cast<llvm::StructType>(structType->getTypeAtIndex(i))));
+            }
+        }
+        else {
+            // Assume the function returns a single memref
+            inferredTypes.push_back(inferSingleMemRef(structType));
+        }
+        return success();
+    }
+    return failure();
+}
+
+LogicalResult runLowering(const CompilerOptions &options, ModuleOp moduleOp,
+                          CompilerOutput::PipelineOutputs &outputs)
+{
+    auto pm = PassManager::on<ModuleOp>(options.ctx, PassManager::Nesting::Implicit);
+
+    std::unordered_map<void *, std::list<Pipeline::Name>> pipelineTailMarkers;
```

Comment:
I would like an invariant to hold here, saying like "users should always be able to query the output of any pipeline". Now, since we allow empty pipelines, there could be passes that terminate more than one pipeline, hence the solution.

Reply:
Sure. Since it seems like the use case is to support empty pipelines rather than multiple prints per pipeline, a `std::optional` may be more appropriate because there will either be 0 or 1 tail marker, never 2 or more.

Reply:
While it is true that the pipeline contains either zero or one tail pass, it is also true that one pass may terminate many pipelines. Since MLIR only reveals pass arguments to our callback, I think we are restricted to the latter form of the relation.

Reply:
I think we're talking about two different matters here: the type of the key and the type of the value. Using `std::optional` is strictly for the value instead of using `std::list` because there can be 0 or 1 tail pass but not more, independently of the key.

On the key side, does using a pointer to the pass correctly de-duplicate when the same pass terminates multiple pipelines? If so, I wonder if we could use a `Pass *` or even a `PassManager::pass_iterator` as the key. It strikes me as more readable than using `void *`.

Reply:
I mean, for configuration like `[('A',['pass1']),('B',[]),('C',[]),('D',['pass2'])]`,  `pass1` is terminating for 3 pipelines: A, B and C. MLIR communicates to us the pass pointer, so I used a list be able to refer to all these names in the correct order.

Regarding the `void*` type - the idea was to express the fact that we are not going to dereference these pointers. But I see that it look unusual, so I am ready to change it to `Pass*` for better readability. I would prefer `Pass*` to iterator because it is the exact type we get from the callback. (will make this change soon)

Reply:
Ah I see, what I had in mind for this case was to not print anything at the end of pipelines B and C because they're empty. If we want to make sure that we print something at the end of each pipeline, then I see the reason for storing a list.

> Regarding the `void*` type - the idea was to express the fact that we are not going to dereference these pointers. But I see that it does look unusual, so I am ready to change it to `Pass*` for better readability. I would prefer `Pass*` to an iterator type because it is the exact type we get from the callback. (will make this change soon)

Sounds great, thanks! Feel free to resolve this conversation after the readability change.